### PR TITLE
add try / except to address issue #10154 to_datetime, Inconsistent be…

### LINKED ISSUE #10154: to_datetime, inconsistent behavior with invalid dates.
Consider Feb 29 1991 (2291991) and March 32 1991 (3321991), Both are invalid dates.
```
pd.to_datetime(2291991, format="%m%d%Y", coerce=True, exact=True)
```
Returns a TypeError: ValueError: day is out of range for month

While
```
pd.to_datetime(3321991, format="%m%d%Y", coerce=True, exact=True)
```
Returns NaT. Which is what I would expect.

In any case they should return the same error or value

Joris Van den Bossche pointed out on the mailing list:
"It has something to do with the number of the day, only values above 31 convert to NaT, 31 or lower raises the error (eg also 31 April raise error instead of giving NaT)"

### PR DESCRIPTION
**Pull request by vincentdavis**
closes #10154 to_datetime, Inconsistent behavior with invalid dates.

results are now

```
In [2]: pd.to_datetime('2015-02-29', format="%Y-%m-%d", coerce=True)
Out[2]: NaT

In [3]: pd.to_datetime('2015-03-32', format="%Y-%m-%d", coerce=True)
Out[3]: NaT

In [4]: pd.to_datetime('2015-02-32', format="%Y-%m-%d", coerce=True)
Out[4]: NaT

In [5]: pd.to_datetime('2015-04-31', format="%Y-%m-%d", coerce=True)
Out[5]: NaT
```

### CODE REVIEW

**Comment by jreback** [comment]
needs tests

File: pandas/tests/test_tseries.py [hunk 1]
```diff
@@ -737,6 +737,13 @@ def test_get_period_field_raises_on_out_of_range(self):
     def test_get_period_field_array_raises_on_out_of_range(self):
         self.assertRaises(ValueError, period.get_period_field_arr, -1, np.empty(1), 0)
 
+class TestDaysInMonth(tm.TestCase):
+    def test_day_not_in_month_coerce_true(self):
```
Comment (jreback) [inline pandas/tests/test_tseries.py]: pls add the issue number as a comment

File: pandas/tests/test_tseries.py [hunk 2]
```diff
@@ -737,6 +737,13 @@ def test_get_period_field_raises_on_out_of_range(self):
...
+        self.assertTrue(isnull(to_datetime('2015-02-29', coerce=True)))
+        self.assertTrue(isnull(to_datetime('2015-02-29', format="%Y-%m-%d", coerce=True)))
+        self.assertTrue(isnull(to_datetime('2015-02-32', format="%Y-%m-%d", coerce=True)))
+        self.assertTrue(isnull(to_datetime('2015-04-31', format="%Y-%m-%d", coerce=True)))
+
```
Comment (jreback) [inline pandas/tests/test_tseries.py]: can you add off these cases again, but with ``coerce=False`` (and assert that they raise ``ValueError``)

File: pandas/tests/test_tseries.py (hunk 1 above)
Comment (vincentdavis) [inline pandas/tests/test_tseries.py]: I thought I did, not sure what went wrong.

File: pandas/tests/test_tseries.py [hunk 3]
```diff
@@ -737,6 +737,17 @@ def test_get_period_field_raises_on_out_of_range(self):
...
         self.assertRaises(ValueError, period.get_period_field_arr, -1, np.empty(1), 0)
 
+class TestDaysInMonth(tm.TestCase):
+    def test_day_not_in_month_coerce_true(self):
+        self.assertTrue(isnull(to_datetime('2015-02-29', coerce=True)))
+        self.assertTrue(isnull(to_datetime('2015-02-29', format="%Y-%m-%d", coerce=True)))
+        self.assertTrue(isnull(to_datetime('2015-02-32', format="%Y-%m-%d", coerce=True)))
+        self.assertTrue(isnull(to_datetime('2015-04-31', format="%Y-%m-%d", coerce=True)))
+    def test_day_not_in_month_coerce_false(self):
+        self.assertRaises(ValueError, to_datetime, '2015-02-29', format="%Y-%m-%d", coerce=False)
```
Comment (jreback) [inline pandas/tests/test_tseries.py]: yeh, test both ``errors='ignore'`` and ``errors='raise'`` here

Comment (vincentdavis) [inline pandas/tests/test_tseries.py]: This looks correct
```
In [10]: pd.to_datetime('2015-02-29', errors='ignore',  coerce=False)
Out[10]: '2015-02-29'
```
I would expect the same from this but get a ValueError.
```
In [12]: pd.to_datetime('2015-02-29', errors='ignore', format="%Y-%m-%d", coerce=False)
....
ValueError: day is out of range for month
```
adding format="%Y-%m-%d" should not change the output.
Is this correct and also needs to be fixed?

**Comment by jreback** [comment]
can you rebase

**Comment by vincentdavis** [comment]
On Fri, Jun 26, 2015 at 5:39 PM, jreback <notifications@github.com> wrote:

> can you rebase


I will work on that this weekend.


Vincent Davis
720-301-3003

**Pull request closed without merging**